File: server/src/controller/payment.controller.ts
import { Controller, Post, Get, Body, Inject, Param } from "@midwayjs/core";
import { Context } from "@midwayjs/koa";
import { ApiTags, ApiOperation } from "@midwayjs/swagger";
import { PaymentService } from "../service/payment.service";
import { ResponseUtil } from "../common/types/response";
import { IUserContext } from "../interface";

@ApiTags("payment")
@Controller("/api/payments")
export class PaymentController {
  @Inject()
  paymentService!: PaymentService;

  @Post("/create")
  @ApiOperation({ summary: "创建支付单" })
  async createPayment(
    ctx: Context,
    @Body() body: { orderId: number; method: string },
  ) {
    const user = ctx.state.user as IUserContext;
    const payment = await this.paymentService.createPayment(
      body.orderId,
      body.method,
    );
    return ResponseUtil.success(payment, "支付单已创建");
  }

  @Post("/callback")
  @ApiOperation({ summary: "支付回调（模拟）" })
  async payCallback(@Body() body: { orderId: number; tradeNo: string }) {
    const payment = await this.paymentService.handlePaySuccess(
      body.orderId,
      body.tradeNo,
    );
    return ResponseUtil.success(payment, "支付成功");
  }

  @Post("/refund")
  @ApiOperation({ summary: "申请退款" })
  async refund(
    ctx: Context,
    @Body() body: { orderId: number; amount: number; reason: string },
  ) {
    const user = ctx.state.user as IUserContext;
    const refund = await this.paymentService.refund(
      body.orderId,
      body.amount,
      body.reason,
    );
    return ResponseUtil.success(refund, "退款申请已提交");
  }

  @Get("/:orderId")
  @ApiOperation({ summary: "查询支付状态" })
  async getPayment(@Param("orderId") orderId: string) {
    const payment = await this.paymentService.findByOrderId(
      parseInt(orderId, 10),
    );
    if (!payment) return ResponseUtil.error("支付单不存在", 404);
    return ResponseUtil.success(payment);
  }
}

File: server-java/src/main/java/com/wudong/controller/PaymentController.java
package com.wudong.controller;

import com.wudong.common.response.ApiResponse;
import com.wudong.dto.payment.ApplyRefundRequest;
import com.wudong.dto.payment.CreatePaymentRequest;
import com.wudong.entity.Payment;
import com.wudong.service.PaymentService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@Tag(name = "支付")
@RestController
@RequestMapping("/api/payments")
@RequiredArgsConstructor
public class PaymentController {

    private final PaymentService paymentService;

    @Operation(summary = "创建预支付")
    @PostMapping("/create")
    public ApiResponse<Payment> createPayment(@Valid @RequestBody CreatePaymentRequest request) {
        Payment payment = paymentService.createPayment(request.getOrderId(), request.getMethod());
        return ApiResponse.success(payment);
    }

    @Operation(summary = "支付成功回调")
    @PostMapping("/callback")
    public ApiResponse<Void> payCallback(@RequestBody Map<String, String> body) {
        paymentService.handlePaySuccess(
                Long.valueOf(body.get("orderId")),
                body.get("tradeNo"),
                body.get("method"));
        return ApiResponse.success(null, "支付成功");
    }

    @Operation(summary = "申请退款")
    @PostMapping("/refund")
    public ApiResponse<Void> applyRefund(@Valid @RequestBody ApplyRefundRequest request) {
        paymentService.applyRefund(request.getOrderId(), request.getAmount(), request.getReason());
        return ApiResponse.success(null, "退款申请已提交");
    }

    @Operation(summary = "查询支付状态")
    @GetMapping("/{orderId}")
    public ApiResponse<Payment> getPayment(@PathVariable Long orderId) {
        return ApiResponse.success(paymentService.getPaymentByOrderId(orderId));
    }
}
